gr-digital/python/digital/qa_decision_feedback_equalizer.py

```diff
@@ -0,0 +1,157 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+#
+# Copyright 2019, 2020 Perspecta Labs Inc.
+#
+# This is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 3, or (at your option)
+# any later version.
+#
+# This software is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this software; see the file COPYING.  If not, write to
+# the Free Software Foundation, Inc., 51 Franklin Street,
+# Boston, MA 02110-1301, USA.
+#
+
+from gnuradio import gr, gr_unittest
+
+import random, numpy
+from gnuradio import digital, blocks, channels
+
+class qa_linear_equalizer(gr_unittest.TestCase):
+
+    def unpack_values(self, values_in, bits_per_value, bits_per_symbol):   
+        # verify that 8 is divisible by bits_per_symbol 
+        m = bits_per_value / bits_per_symbol
+        # print(m)
+        mask = 2**(bits_per_symbol)-1
+            
+        if bits_per_value != m*bits_per_symbol:
+            print("error - bits per symbols must fit nicely into bits_per_value bit values")
+            return []
+            
+        num_values = len(values_in)
+        num_symbols = int(num_values*( m) )
+        
+        cur_byte = 0
+        cur_bit = 0
+        out = []
+        for i in range(num_symbols):
+            s = (values_in[cur_byte] >> (bits_per_value-bits_per_symbol-cur_bit)) & mask
+            out.append(s)
+            cur_bit += bits_per_symbol
+            
+            if cur_bit >= bits_per_value:
+                cur_bit = 0
+                cur_byte += 1
+                
+        return out
+
+    def map_symbols_to_constellation(self, symbols, cons):
+        l = list(map(lambda x: cons.points()[x], symbols))
+        return l
+
+
+    def setUp(self):
+        random.seed(987654)
```

Comment:
This likely won't ever become a problem here, but beware: unit tests are executed in parallel (though at this point in different python interpreters), and thus the state of the global `random` might change unexpectedly! (this is more relevant for python blocks, though.) 

A way around this could be `rng = gr.random(987654);` here and and instead of getrandbits `rng.ran_int() & 0xFF` or so below :) (MEMO TO SELF: might want to switch `ran_int` from std:: to gr::xoroshiro at some point for speed reasons)